### PR TITLE
node: Represent workspace submodules as Projects

### PR DESCRIPTION
**Pull request by fviernau**
See individual commits for the details.

Fixes https://github.com/oss-review-toolkit/ort/issues/9196, fixes https://github.com/oss-review-toolkit/ort/issues/8940.

**Important**: Previously, submodules were represented as packages which were not referenced by any project scope. 
                             As consequence, this could lead to incorrect handling of all their transitive dependencies by the policy rules.

### CODE REVIEW

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml [hunk 1]
```diff
@@ -28,6 +28,42 @@ analyzer:
...
+        type: "Git"
+        url: "<REPLACE_URL_PROCESSED>"
+        revision: "<REPLACE_REVISION>"
+        path: "plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces/src/app"
+      homepage_url: ""
+      scopes:
+      - name: "dependencies"
+        dependencies:
+        - id: "PNPM::testing-pnpm-package-a:1.0.2"
+          linkage: "PROJECT_DYNAMIC"
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml:54]: Good question. How do we handle it e.g. for Pub with references to Gradle projects, do we there also list the Gradle projects separately, @mnonnenmacher?

In any case, at least in the dependency graph representation the duplication should not matter much memory-wise, as it should be dedupli9cated.

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml [hunk 2]
```diff
@@ -28,6 +28,42 @@ analyzer:
     skip_excluded: false
   result:
     projects:
+    - id: "PNPM::pnpm-app-example:1.1.4"
```
[the remark this answers is not in the document]
Comment (fviernau) [reply plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml:31]: Yes, good catch.

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml:54]: If we decide to do this, I believe it'd be good to do this in a separate commit anyway, to have it properly documented and to not increase this commit further. So, I propose we make the discussion non-blocking for this PR, and I will make a follow-up PR if needed.

File: plugins/package-managers/node/src/main/kotlin/Npm.kt [hunk 3]
```diff
@@ -184,79 +182,65 @@ open class Npm(
...
 
-        // Create packages for all modules found in the workspace and add them to the graph builder. They are
-        // reused when they are referenced by scope dependencies.
-        val isWorkspacesProject = (findWorkspaceSubmodules(workingDir) - workingDir).isNotEmpty()
-        if (isWorkspacesProject) {
-            val packages = parseInstalledModules(workingDir)
-            graphBuilder.addPackages(packages)
-        }
+        return projectDirs.map { projectDir ->
+            val issues = mutableListOf(*installIssues.takeIf { projectDir == workingDir }.orEmpty().toTypedArray())
```
[the remark this answers is not in the document]
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/Npm.kt:201]: I tried to make it less cryptic by introducing an `if` condition on a separate line.

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml [hunk 4]
```diff
@@ -50,16 +86,31 @@ analyzer:
       scopes:
       - name: "dependencies"
         dependencies:
-        - id: "NPM::chalk:4.0.0"
+        - id: "NPM::json-stable-stringify:1.0.1"
```
Comment (sschuberth) [inline plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml:89]: Shouldn't `PNPM::pnpm-workspaces:1.0.1` also have `PROJECT_DYNAMIC` dependencies on the projects referred to by its workspces, i.e. on `PNPM::pnpm-app-example:1.1.4`, `PNPM::testing-pnpm-package-a:1.0.2` and `PNPM::testing-pnpm-package-b:1.0.2`?

Comment (sschuberth) [reply plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml:89]: I'm actually not sure about this myself. Maybe a workspace declaration odes not really express a dependency relationship by itself, and thus these projects should not be listed here.

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/yarn/workspaces-expected-output.yml [hunk 5]
```diff
@@ -15,14 +15,52 @@ project:
...
+    url: ""
+    revision: ""
+    path: ""
+  vcs_processed:
+    type: "Git"
+    url: "<REPLACE_URL_PROCESSED>"
+    revision: "<REPLACE_REVISION>"
+    path: "plugins/package-managers/node/src/funTest/assets/projects/synthetic/yarn/workspaces"
+  homepage_url: ""
+  scopes: []
```
Comment (sschuberth) [inline plugins/package-managers/node/src/funTest/assets/projects/synthetic/yarn/workspaces-expected-output.yml:43]: Continuing on my earlier question, what do you think about having a follow-up PR that creates a scope called "workspace" which *does* list the projects that belong to a workspace as `PROJECT_DYNAMIC` dependencies?

**Review by sschuberth — COMMENTED** [review]
Do you know why no changes to the expected results for NPM and Yarn2 workspaces were necessary?

**Comment by fviernau** [comment]
> Do you know why no changes to the expected results for NPM and Yarn2 workspaces were necessary?

For NPM we never implemented support for workspaces / we do not support it. Yarn2 is a separate implementation, it luckily does not also inherit from NPM.

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml (hunk 4 above)
Comment (fviernau) [reply plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/workspaces-expected-output.yml:89]: I planned to investigate this in my following work. In pnpm I noticed that a workspace actually can have dependencies, but workspace submodules are not shown under dependencies. To me this makes sense. 
I believe a workspace is just some construct which helps with development, and also helps with using a single lockfile for all projects, but it does not really matter in terms of the actual dependencies. 

I propose to not add any such dependency from the workspace to the submodules. At least not in this PR. I guess there will be a better understanding on this on my side soon anyway, so if you're fine, let's defer the topic.

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/yarn/workspaces-expected-output.yml (hunk 5 above)
Comment (fviernau) [reply plugins/package-managers/node/src/funTest/assets/projects/synthetic/yarn/workspaces-expected-output.yml:43]: I'd tend to not do that, however, similar to my answer above, I'd propose to defer this decision.

**Review by sschuberth — APPROVED** [review]

**Comment by sschuberth** [comment]
> > For NPM we never implemented support for workspaces / we do not support it.
> 
> Oh! I was mislead by us having an [`npm/workspaces` directory](https://github.com/oss-review-toolkit/ort/tree/main/plugins/package-managers/node/src/funTest/assets/projects/synthetic/npm/workspaces), but indeed it's not being used from `NpmFunTest`...

Actually, are you sure about that @fviernau? At least the NPM code has stuff like

https://github.com/oss-review-toolkit/ort/blob/864d19f5b8669605c2d80729c51c7469c67250d0/plugins/package-managers/node/src/main/kotlin/Npm.kt#L127-L148

**Comment by sschuberth** [comment]
Maybe I just forgot to add an actual test for `plugins/package-managers/node/src/funTest/assets/projects/synthetic/npm/workspaces` in f99e2ede5053200cef966836b051e77f5328d575, hmm...

**Comment by fviernau** [comment]
> Actually, are you sure about that @fviernau? At least the NPM code has stuff like

If found that quite a bit of logic has been introduced into NPM which at the time of introduction had only been used by one child class. So, I wouldn't assume per default that code in NPM is actually used for NPM.

**Comment by fviernau** [comment]
> Maybe I just forgot to add an actual test for `plugins/package-managers/node/src/funTest/assets/projects/synthetic/npm/workspaces` in [f99e2ed](https://github.com/oss-review-toolkit/ort/commit/f99e2ede5053200cef966836b051e77f5328d575), hmm...

I saw this commit as well when working on this PR. I though that it was only about the detection, e.g. decide which package manager to use for a particular project dir / definition file.

**Comment by fviernau** [comment]
Note: Adding NPM workspace support was on my TODO list, and I plan to do this only after separating the package managers from one another.

**Comment by sschuberth** [comment]
> I though that it was only about the detection, e.g. decide which package manager to use for a particular project dir / definition file.

Could very well be, I currently do not recall the details anymore.

**Pull request merged**